pkg/controller/config/config_controller.go

```diff
@@ -0,0 +1,305 @@
+	}
+
+	// Generate render data.
+	renderData, err := GenerateRenderData(operConfig)
+	if err != nil {
+		log.Error(err, "failed to generate render data")
+		r.status.SetDegraded(statusmanager.OperatorConfig, "RenderConfigError",
+			fmt.Sprintf("Failed to render operator configurations: %v", err))
+		return reconcile.Result{}, err
+	}
+
+	// Compare configurations change.
+	appliedConfig, err := r.getAppliedOperConfig()
+	if err != nil {
+		log.Error(err, "failed to get applied config")
+		r.status.SetDegraded(statusmanager.OperatorConfig, "InternalError",
+			fmt.Sprintf("Failed to get current configurations: %v", err))
+		return reconcile.Result{}, err
+	}
+	agentNeedChange, controllerNeedChange := NeedApplyChange(appliedConfig, operConfig)
+	if !agentNeedChange && !controllerNeedChange {
+		log.Info("no configuration change, skip update")
+		return reconcile.Result{}, nil
+	}
+
+	// Render configurations.
+	objs, err := render.RenderDir(operatortypes.DefaultManifestDir, renderData)
+	if err != nil {
+		log.Error(err, "failed to render configuration")
+		r.status.SetDegraded(statusmanager.OperatorConfig, "RenderConfigError",
+			fmt.Sprintf("Failed to render operator configurations: %v", err))
+		return reconcile.Result{}, err
+	}
+
+	// Apply configurations.
+	for _, obj := range objs {
+		if err = controllerutil.SetControllerReference(clusterConfig, obj, r.scheme); err != nil {
+			log.Error(err, "failed to set owner reference")
+			r.status.SetDegraded(statusmanager.OperatorConfig, "ApplyObjectsError",
+				fmt.Sprintf("Failed to set owner reference: %v", err))
+			return reconcile.Result{}, err
+		}
+		if err = apply.ApplyObject(context.TODO(), r.client, obj); err != nil {
+			log.Error(err, "failed to apply resource")
+			r.status.SetDegraded(statusmanager.OperatorConfig, "ApplyObjectsError",
+				fmt.Sprintf("Failed to apply operator configurations: %v", err))
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Delete old antrea-agent and antrea-controller pods.
+	if r.appliedOperConfig != nil && agentNeedChange {
+		if err = deleteExistingPods(r.client, operatortypes.AntreaAgentDaemonSetName); err != nil {
+			r.status.SetDegraded(statusmanager.OperatorConfig, "DeleteOldPodsError",
+				fmt.Sprintf("DaemonSet %s is not using the latest configuration updates because: %v",
+					operatortypes.AntreaAgentDaemonSetName, err))
+			return reconcile.Result{}, err
+		}
+	}
+	if r.appliedOperConfig != nil && controllerNeedChange {
+		if err = deleteExistingPods(r.client, operatortypes.AntreaControllerDeploymentName); err != nil {
+			r.status.SetDegraded(statusmanager.OperatorConfig, "DeleteOldPodsError",
+				fmt.Sprintf("Deployment %s is not using the latest configuration updates because: %v",
+					operatortypes.AntreaControllerDeploymentName, err))
+			return reconcile.Result{}, err
+		}
+	}
+	// Update cluster network CR status.
+	clusterNetworkConfigChanged := HasClusterNetworkConfigChange(r.appliedClusterConfig, clusterConfig)
```

Comment:
For operator restart case, appliedClusterConfig is nil and operator will update network CRD status in this case. But it should be fine as no impact for other resources

Reply:
Yes, it's hard to get old config in current controller framework. We can enhance it later.

Reply:
Please make sure to add all TODO in code in case we forget them later.

Reply:
Done.